### ISSUE TITLE
mlflow dataset

### ISSUE THREAD

**Issue by luizvbo** [kedro-org/kedro-plugins#248]
## Description
When using mlflow with Kedro, it's currently not possible to connect the node producing the model to the one loading it (e.g., for validation). This could be solved by adding a dataset to handle mlflow `load_model`, `save_model` and `log_model` operations. 

## Context
By having a dataset for mlflow, we can take advantage of the model versioning control provided by mlflow while adequately representing the model saving and loading operations in the Kedro DAG.

## Possible Implementation
The dataset should provide the option of saving the model locally (`save_model`) and using a mlflow instance (`log_model`). For the latter, Kedro needs to access the mlflow session.

**Comment by datajoely** [comment]
Have you considered using the `kedro-mlflow` plug-in which has some of these already?
https://kedro-mlflow.readthedocs.io/en/stable/source/07_python_objects/index.html

**Comment by luizvbo** [comment]
Yes, I came across the plugin, but I wanted to keep my list of requirements as small as possible (and kedro-datasets is already part of the list). Would you be open to integrating the dataset into the repo in case I implement it?

**Comment by astrojuanlu** [comment]
Hi @luizvbo , could you describe in a bit more detail if Kedro-mlflow suits your needs, or if you’re looking for different functionality somehow? Also it’s not clear to me whether you have already implemented such a dataset, or if you’re requesting that someone else does it.

**Comment by luizvbo** [comment]
Hi @astrojuanlu! My current use case is slightly different from the one from Kedro-mlflow. 
- In my use case, I would like to use mlflow to store a model trained in one of my Kedro nodes and load it back in a subsequent node (e.g., for model validation). 
  -  mlflow provides different methods to store models according to their type (e.g., pytorch, sklearn, spark, spacy, tensorflow), which can be used accordingly by a new Kedro dataset.
  - Using an mlflow dataset wouldn't (necessarily) involve tracking all the nodes using mlflow. Hence, the Kedro-mlflow plugin would be overkill for this scenario. 
- In Kedro-mlflow, the dataset class provided (MlflowArtifactDataSet) wraps a Kedro dataset (from kedro-datasets) and stores the data using `mlflow.log_artifact`. Hence, the object to be stored are handled by the datasets already provided by Kedro-datasets.

I am currently working on dataset class that implements the idea. My question is whether it would add value to have it integrated into Kedro-datasets. I can create a PR with my implementation once it's more mature.

**Comment by SajidAlamQB** [comment]
Hi @luizvbo,

Thanks for the clarifying, this sounds like an interesting and potentially useful feature. I think there could be some value in incorporating your implementation into `Kedro-datasets`, especially for users like yourself who want to take advantage of the model versioning features provided by MLflow but don't require all the functionality of the kedro-mlflow plugin.

I'd suggest creating the new custom dataset in your own project, or even potentially creating a small plugin for it. One it's ready please do share your implementation with us, we would love to see it and provide feedback or help to promote it!

**Comment by astrojuanlu** [comment]
cc @galileo-galilei

**Comment by Galileo-Galilei** [comment]
Hi everyone (disclaimer for @luizvbo : i am the author of ``kedro-mlflow``), just chiming in to give some information about above questions: 
- ``kedro-mlflow`` has [specific datasets to handle models](), including [``MlflowModelSaverDataSet``](https://github.com/Galileo-Galilei/kedro-mlflow/blob/master/kedro_mlflow/io/models/mlflow_model_saver_dataset.py) and  [``MlflowModelLoggerDataSet``](https://github.com/Galileo-Galilei/kedro-mlflow/blob/master/kedro_mlflow/io/models/mlflow_model_logger_dataset.py) which perform exactly what you suggest. In fact they are compatible with any mlflow backend, including (e.g., pytorch, sklearn, spark, spacy, tensorflow and absolutely all the ones declared in [mlflow documentation](https://mlflow.org/docs/latest/models.html#built-in-model-flavors))
- ``kedro-mlflow`` only depends on [kedro, mlflow and pydantic](https://github.com/Galileo-Galilei/kedro-mlflow/blob/master/requirements.txt) so the risk of dependencies conflicts is very low. I you really want to minimize the surface, you can copy the code and import ``kedro`` and ``mlflow`` only (maybe ``mlflow-skinny`` will be enough, but you'll have to check).
- To avoid automatic parameters tracking, you can just [ignore ``kedro-mlflow`` in your ``settings.py``](https://docs.kedro.org/en/stable/hooks/introduction.html#disable-auto-registered-plugins-hooks)
```python
# src/<package_name>/settings.py

DISABLE_HOOKS_FOR_PLUGINS = ("kedro-mlflow",)
```

But if you want to integrate it to the ``kedro-datasets`` implementation or customize the datasets, let's go for it!

**Comment by luizvbo** [comment]
@Galileo-Galilei I hadn't seen the classes you shared. I think they do exactly what I was looking for. In this case, I will use them instead of implementing new ones. Thank you.
